lib/rubygems/precompiler.rb

```diff
@@ -6,16 +6,32 @@
 require 'zlib'
 require 'pathname'
 
+
 class Gem::Precompiler
   include FileUtils
 
   def initialize(gemfile, opts = {})
-    @installer = Gem::Installer.new(gemfile, opts.dup.merge(:unpack => true))
+    @installer = Gem::Installer.new(gemfile, opts.dup.merge(:unpack => true, :build_args => opts.fetch(:build_config,[])))
     @spec = @installer.spec
 
     @target_dir = opts.fetch(:output, Dir.pwd)
     @target_dir = File.join(@target_dir, arch_string) if opts.fetch(:arch, false)
+    @debug = opts.fetch(:debug, false)
     @options = opts
+
+    # This writes out a build_info file that the extension builder will process to set the
+    # build configuration. We use the write_build_info_file method which should work on ruby 2+
+    # on ruby  < 1.9.3 we don't support setting of build options. 1.9.3 is EOL.
+    # However most simple gems that do not require build time config will still work.
+    if !opts.fetch(:build_config,[]).empty?
```

Comment:
I'd be tempted to return early here, there's a fair bit going on inside this nested condition.

```ruby
return if opts.fetch(:build_config, []).empty?

if Gem::Installer.method_defined?…
```